File: 0x00基本算法/习题/AcWing120.cpp
#include <iostream>
#include <algorithm>
using namespace std;
const int N = 200006, INF = 0x3f3f3f3f;
int n, t, s[N], e[N], d[N];
// b-a后以d为间隔，间隔数+1即位点数
int f(int a, int b, int d) {
  return (b >= a) ? ((b - a)/d + 1) : 0;
}
// 求有多少个数<=r，之后可根据奇偶性二分
int le(int r) {
  int ans = 0;
  for (int i=1; i<=n; i++) ans += f(s[i], min(e[i], r), d[i]);
  return ans;
}

void Defense() {
  cin >> n;
  int mins = INF, maxe = 0;
  for (int i=1; i<=n; i++) {
    cin >> s[i] >> e[i] >> d[i];
    mins = min(mins, s[i]);
    maxe = max(maxe, e[i]);
  }
  if (le(maxe) % 2 == 0) {
    cout << "There's no weakness." << endl;
    return;
  }
  int l = mins, r = maxe;
  while (l < r) {
    int mid = (l + r) >> 1;
    if (le(mid) % 2) r = mid;
    else l = mid + 1;
  }
  cout << l << " " << le(l) - le(l-1) << endl;
}

int main() {
  cin >> t;
  while (t--) Defense();
  return 0;
}

File: 0x00基本算法/习题/AcWing120.py
from math import inf
N = 200006
n, t, s, e, d = 0, 0, [0]*N, [0]*N, [0]*N
def f(a, b, c):
    return (b-a)//c + 1 if b>=a else 0
    
def le(r):
    global n
    ans = 0
    for i in range(1, n+1): ans += f(s[i], min(e[i], r), d[i])
    return ans
    
def Defense():
    global n
    n = int(input())
    mins, maxe = inf, 0
    for i in range(1, n+1):
        s[i], e[i], d[i] = map(int, input().split())
        mins = min(mins, s[i])
        maxe = max(maxe, e[i])
    
    if le(maxe) % 2 == 0:
        print("There's no weakness.")
        return
    
    l, r = mins, maxe
    while l < r:
        mid = (l + r) >> 1
        if le(mid) % 2: r = mid
        else: l = mid + 1
    print(l, le(l)-le(l-1))

def main():
    global t
    t = int(input())
    for _ in range(t): Defense()
    
main()